node/src/runner.rs

```diff
@@ -37,36 +40,58 @@ use crate::{
     config::NodeConfig,
     options::{Command, Options, RunOptions},
     regtest_options::ChainConfigOptions,
+    storage_config::StorageBackend,
 };
 
+/// Set up storage and initialize chainstate
+fn make_chainstate<B: 'static + storage::Backend>(
+    storage_backend: B,
+    chain_config: &Arc<ChainConfig>,
+    chainstate_config: chainstate::ChainstateConfig,
+) -> Result<Box<dyn ChainstateInterface>> {
+    let storage = chainstate_storage::Store::new(storage_backend)?;
+    let chainstate = chainstate::make_chainstate(
+        Arc::clone(chain_config),
+        chainstate_config,
+        storage,
+        DefaultTransactionVerificationStrategy::new(),
+        None,
+        Default::default(),
+    )?;
+    Ok(chainstate)
+}
+
 /// Initialize the node, giving caller the opportunity to add more subsystems before start.
 pub async fn initialize(
     chain_config: ChainConfig,
     node_config: NodeConfig,
 ) -> Result<subsystem::Manager> {
     let chain_config = Arc::new(chain_config);
 
-    // Initialize storage.
-    // TODO: Uses in-memory for now, will be configurable once multiple backends are implemented.
-    let storage = chainstate_storage::inmemory::Store::new_empty()?;
-
     // INITIALIZE SUBSYSTEMS
 
     let mut manager = subsystem::Manager::new("mintlayer");
     manager.install_signal_handlers();
 
+    // Pick default storage backend based on chain type
+    let backend_type =
+        node_config.storage.backend.unwrap_or_else(|| match chain_config.chain_type() {
+            ChainType::Signet | ChainType::Mainnet | ChainType::Testnet => StorageBackend::Lmdb,
+            ChainType::Regtest => StorageBackend::InMemory,
+        });
+
     // Chainstate subsystem
-    let chainstate = manager.add_subsystem(
-        "chainstate",
-        chainstate::make_chainstate(
-            Arc::clone(&chain_config),
-            node_config.chainstate,
-            storage.clone(),
-            DefaultTransactionVerificationStrategy::new(),
-            None,
-            Default::default(),
-        )?,
-    );
+    let chainstate = match backend_type {
+        StorageBackend::Lmdb => {
+            let storage_backend = storage_lmdb::Lmdb::new(node_config.datadir);
```

Comment:
Yeah, I don't think this is correct. Passing `node_config.datadir` basically means that the lmdb storage will be directly in datadir, which is not really what we need. There are other things that go into the datadir, such as the p2p noise key. We need a subdirectory for lmdb, say `chainstate-lmdb`. So something like `node_config.datadir.join("chainstate-lmdb")` is what we need.

Reply:
Ok. I was also wondering if it makes sense to incorporate chain type into the name. For example, when the database contains mainnet blocks, and the user launches the node in testnet mode, different database should be used.

Reply:
Typically, the testnet has a different data directory because everything in it is different and shouldn't share anything with the mainnet data directory. It could be a subdirectory in the mainnet directory called testnet/regtest/etc.